src/environmentVariables.ts

```diff
@@ -0,0 +1,147 @@
+import * as os from 'os';
+import * as util from 'util';
+
+const inspect = util.inspect.custom;
+const envProperty = Symbol('envProperty');
+
+export type EnvironmentVariablesBasic = Record<string, string>;
+export type EnvironmentVariablesNull = Record<string, string | null>;
+export type EnvironmentVariablesUndefined = Record<string, string | undefined>; // alias of NodeJS.ProcessEnv
+export type EnvironmentVariables = Record<string, string | undefined | null>;
+export type EnvironmentVariablesItem = EnvironmentVariables | null | undefined;
+
+class EnvironmentVariablesPrivate {
+    private keyMapping: Map<string, string>;
+    public [envProperty]: EnvironmentVariables;
+    protected isWin32: boolean;
+    protected preserveNull: boolean;
+    constructor(
+        preserveNull: boolean | undefined,
+        isWin32: boolean | undefined) {
+        this.keyMapping = new Map<string, string>();
+        this[envProperty] = {};
+        if (preserveNull === undefined) {
+            this.preserveNull = false;
+        } else {
+            this.preserveNull = preserveNull;
+        }
+        if (isWin32 === undefined) {
+            this.isWin32 = os.platform() === 'win32';
+        } else {
+            this.isWin32 = isWin32;
+        }
+    }
+
+    public getKey(key: string, updateKey: boolean): string {
+        if (this.isWin32) {
+            const normalizedKey = key.toUpperCase();
+            let resultKey = this.keyMapping.get(normalizedKey);
+            if (resultKey === undefined) {
+                resultKey = key;
+                if (updateKey) {
+                    this.keyMapping.set(normalizedKey, resultKey);
+                }
+            }
+            return resultKey;
+        }
+        return key;
+    }
+
+    public get(key: string | symbol): string | undefined | null {
+        if (typeof key === 'string') {
+            return this[envProperty][this.getKey(key, false)];
+        }
+        return undefined;
+    }
+
+    public set(key: string | symbol, value?: string | null, receiver?: any): boolean {
+        if (typeof key === 'string') {
+            let needSet = true;
+            if (value === undefined) {
+                needSet = false;
+            } else if (value === null) {
+                if (!this.preserveNull) {
+                    needSet = false;
+                }
+            } else if (typeof value !== 'string') {
+                value = '' + value;
+            }
+            if (needSet) {
+                const existKey = this.getKey(key, true);
+                return Reflect.set(this[envProperty], existKey, value, receiver);
+            } else {
+                return Reflect.deleteProperty(this[envProperty], key);
+            }
+        }
+        return false;
+    }
+
+    [inspect]() {
+        return util.inspect(this[envProperty]);
+    }
+
+    toString() {
+        return this[envProperty].toString();
+    }
+}
+
+export class EnvironmentVariablesUtils {
```

Comment:
It would be really nice if `create` and `merge` could be on `Environment` as static methods.  Because of the proxy, we can't do this right now. If we can't remove the proxy, then please rename this `EnvironmentUtils` instead.